pam_motd: Add support for specifying multiple motd directories

rfairley opened the pull request:
This adds to `pam_motd.c` functions to parse a string of colon-separated directories as arguments for `motd` and `motd_dir`, and check for overriding filenames before displaying the motds. This also displays the motds in lexicographic order of the filenames, after merging the files contained in each directory.

Tests are added to verify the previous single-directory functionality works, as well as with multiple directories, and with symlinks.

If this is suitable to have in PAM, I will also update the documentation in this PR.

Appreciate your review and feedback!

Closes #68

rfairley commented:
Pushed another commit to fix the logic behind displaying the files in lexicographic order - was not merging filenames from directories before sorting before. Will squash the commit afterwards.

lucab commented on modules/pam_motd/pam_motd.c:
```diff
@@ -97,9 +98,180 @@ static void try_to_display_directory(pam_handle_t *pamh, const char *dirname)
+    for (int i = 0; i < num_motd_dirs; i++) {
+	for (int j = 0; j < namelist_sizes[i]; j++) {
+	   namelist_final[num_files] = namelist[i][j]->d_name;
+	num_files++;
```
Misindented line.

rfairley replied:
Addressed

lucab commented on modules/pam_motd/pam_motd.c:
```diff
@@ -97,9 +98,180 @@ static void try_to_display_directory(pam_handle_t *pamh, const char *dirname)
+    int i = 0;
+
+    arg_split = (char **)malloc(sizeof(char *) * num_strs);
+    if (!arg_split) {
```
Here and everywhere else, I think it would be better to have null-pointer checks being explicit (i.e. `!= NULL`).

rfairley replied:
Agreed, making them explicit

lucab commented on modules/pam_motd/pam_motd.c:
```diff
@@ -97,9 +98,180 @@ static void try_to_display_directory(pam_handle_t *pamh, const char *dirname)
+	return NULL;
+    }
+
+    arg_extracted = (char *)strtok(arg, delim);
```
I think the returned pointer cast is not needed here, right?

rfairley replied:
:+1: thanks for catching, removing the cast

rfairley commented on xtests/tst-pam_motd3.sh:
```diff
@@ -0,0 +1,46 @@
+echo "motd: test-from-symlink - not show" > ${TST_DIR}/test-from-symlink.motd
+ln -s ${TST_DIR}/test-from-symlink.motd ${TST_DIR}/run/motd.d/hidden-by-symlink.motd
+
+# Test hidden by a null symlink
```
Right, previously this was done implicitly, but by considering *any* symlink to override a file and following the symlink. It probably is better to check the symlink target for `/dev/null` and not try to print it - working this in

rfairley replied:
Actually - leaving this done implicitly - think it is fine. This is to avoid an extra check like `readlink(2)` being necessary before displaying the motd files

t8m replied:
Having it done implicitly is OK.

rfairley commented on modules/pam_motd/pam_motd.c:
```diff
@@ -97,9 +98,180 @@ static void try_to_display_directory(pam_handle_t *pamh, const char *dirname)
+    arg_extracted = (char *)strtok(arg, delim);
+    while (arg_extracted && i < num_strs) {
+	arg_split[i++] = arg_extracted;
+	arg_extracted = (char *)strtok(NULL, delim);
```
:+1: addressed this

rfairley commented:
Pushed changes from review comments aside from symlink checking `/dev/null` - tst-pam_motd3.sh fails as the symlink logic needs to be fixed. Working on this

rfairley commented:
Pushed changes to fix the problems with symlinks. One problem was forgetting to `strcmp` when checking for duplicates in `namelist_final`, and the other in the test where the symlink should be relative within the testing directory.

t8m commented on modules/pam_motd/pam_motd.c:
```diff
@@ -97,9 +97,192 @@ static void try_to_display_directory(pam_handle_t *pamh, const char *dirname)
+	return;
+    }
+
+    struct dirent ***namelist = NULL;
```
Call namelist better - something like dirlists and dirlists_sizes could be more clear.

rfairley replied:
Addressed - sorry `namelist` was hastily chosen from reading the `dirscan` manpage example, should have thought a bit more about it. Changed to `dirscans` and `dirscans_sizes`, holding the output of a `dirscan`. How does this sound? `dirlists` is also good with me

t8m commented on modules/pam_motd/pam_motd.c:
```diff
@@ -97,9 +97,192 @@ static void try_to_display_directory(pam_handle_t *pamh, const char *dirname)
+
+    _pam_drop(namelist_final);
+    _pam_drop(namelist);
+}
```
You're leaking here members of namelist array, the individual dirent entries, and also the namelist_sizes array.

rfairley replied:
Sorry I missed that - addressed now.

t8m commented on xtests/tst-pam_motd2.c:
```diff
@@ -0,0 +1,65 @@
+/*
```
To avoid code duplication, please just have a single tst-pam_motd.c which would take the service name as parameter.

rfairley replied:
I did this and created a test file `tst-pam_motd.sh`, which then calls the individual `tst-pam_motd{1,2,3}.sh` which each specify the service name. Is this alright to do from the perspective of the testing structure? I can have all the tests done in the one `tst-pam_motd.sh` file as well

t8m replied:
Do as you prefer.

dustymabe commented:
wow. Thanks so much @t8m for the review! It sounds like most of your comments are for code style or implementation detail related. Does this mean if we resolve the minor issues then you'd be +1 for accepting this change ?

t8m commented:
Yes, functionality-wise I think this is OK.

rfairley commented:
Alright, addressed the changes, added the documentation, and think it is ready for another look! Whenever you have a moment of course. Thanks!

rfairley commented on modules/pam_motd/pam_motd.c:
```diff
@@ -323,43 +362,43 @@ int pam_sm_open_session(pam_handle_t *pamh, int flags,
-
     if (motd_path_split != NULL) {
-	for (int i = 0; i < num_motd_paths; i++) {
+	int i = 0;
```
Addressed - initialization removed from here and done only at the for loop.

rfairley commented on modules/pam_motd/pam_motd.c:
```diff
@@ -97,12 +97,234 @@ static void try_to_display_directory(pam_handle_t *pamh, const char *dirname)
+ * and the number of strings OUT_NUM_STRS.
+ * Returns 0 in case of error, 1 in case of success.
+ */
+int pam_split_string(const pam_handle_t *pamh, char *arg, char delim,
```
Addressed

rfairley commented:
Addressed the two changes, with the patch in response to those changes here https://github.com/rfairley/linux-pam/commit/042eb63910a4abc3bdf0e0e29006225c54d8710b. Rebased on current master.

Good to merge from my end. Thanks!

t8m commented:
There is no release scheduled yet so yes, it would be fine to open PR against Fedora package.

asavah commented:
Was this runtime tested?
I'm suspicious this is causing segfault in libpam.so when logging in via sshd making sshd unusable.
I have a pet OS which uses latest and greatest (when possible) packages.
```
Nov 21 04:14:26 e45m1 systemd-coredump[386]: Process 383 (sshd) of user 0 dumped core.

                                             Stack trace of thread 383:
                                             #0  0x00007f54c112b106 __GI___strlen_sse2 (libc.so.6)
                                             #1  0x00007f54c111b15e __GI___strdup (libc.so.6)
                                             #2  0x00007f54c0ff55ba pam_sm_open_session (pam_motd.so)
                                             #3  0x00007f54c15af2e5 n/a (libpam.so.0)
                                             #4  0x000000000040a35b n/a (sshd)
                                             #5  0x00007f54c10bb02e __libc_start_main (libc.so.6)
                                             #6  0x000000000040a50a n/a (sshd)
```
linux-pam built  at https://github.com/linux-pam/linux-pam/commit/f7abb8c1ef3aa31e6c2564a8aaf69683a77c2016 behaves just fine.
x86_64, gcc 8.2, glibc 2.28 (git 2.28 branch)

rfairley commented:
Sorry this broke on your end. Thank you for reporting it. Looks like this is caused by a case where the line specifying the module in `/etc/pam.d/sshd` looks as follows:

```
session optional    pam_motd.so    motd=...
```

(with no `motd_dir` argument given), which makes the `motd_dir_path` pointer equal to `NULL` which is then passed into `strdup`, causing the segfault.

This is something I should have caught while writing, or I should have had a test with these conditions. The tests I added did not cover this. Pushed a patch which fixes this and adds a test case ( #76).

@t8m I opened a new PR for the fix, but can squash the fix commit into this PR if preferred.

dustymabe commented:
@asavah thanks for reporting that. Does the fix in #76 (already merged) work for you?

asavah commented:
@dustymabe Yes, thanks, ssh login works now.

@rfairley No worries, bugs happen ) , good thing that it's fixed now, thanks.
Just FYI this is my system-login which triggered the the crash:
```
cat /etc/pam.d/system-login
#%PAM-1.0

auth       required   pam_tally.so         onerr=succeed file=/var/log/faillog
auth       required   pam_shells.so
auth       requisite  pam_nologin.so
auth       include    system-auth

account    required   pam_access.so
account    required   pam_nologin.so
account    include    system-auth

password   include    system-auth

session    optional   pam_loginuid.so
session    include    system-auth
session    optional   pam_motd.so          motd=/etc/motd
session    optional   pam_mail.so          dir=/var/spool/mail standard quiet
-session   optional   pam_systemd.so
session    required   pam_env.so
```
/etc/motd is a file with one line in it. It works fine now.
